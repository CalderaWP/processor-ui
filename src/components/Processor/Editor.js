import React from 'react';
import PropTypes from 'prop-types';

export class Editor extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			opacity: 1,
			hovered: false,
		};
		this.mouseEnter = this.mouseEnter.bind(this);
		this.mouseLeave = this.mouseLeave.bind(this);
		this.changeType = this.changeType.bind(this);
		this.handleChange = this.handleChange.bind(this);

	}

	/**
	 * Handler for the mouse enter event
	 */
	mouseEnter() {
		this.setState({hovered: true})
	}

	/**
	 * Handler for the mouse leave event
	 */
	mouseLeave() {
		this.setState({hovered: false})
	}

	configFields(){

	}

	handleChange(fieldId,newValue){
		console.log(fieldId,newValue);
	}

	changeType(event){
		this.props.onUpdateProcessor({
			...this.props.processor,
			type: event.target.value
		})
	}


	render() {
		let removeStyle = {
			opacity: 0.1
		};
		if ( this.state.hovered){
			removeStyle.opacity = 1
		}
		const labelAs = this.props.label ? this.props.label : this.props.ID;
		const options = [
			{
				label: 'Email',
				value: 'email'
			},
			{
				label: 'Redirect',
				value: 'redirect'
			}

		];
		return (
			<div
				className={this.props.className}
				onMouseEnter={this.mouseEnter}
				onMouseLeave={this.mouseLeave}
			>
				<div>{labelAs}</div>
				<div>
					Type: <span className={'processor-type'}>{this.props.processor.type}</span>
				</div>
				{ ! this.props.processor.type &&
					<select
						className={'processor-type-chooser'}
						onChange={this.changeType}
						value={this.props.type}
					>
						<option>---</option>
						{options.map(option => {
								return (
									<option
										key={option.value}
										value={option.value}
									>
										{option.label}
									</option>
								);

							}
						)}
					</select>
				}
				{this.props.processor.configFields &&
					<div
						className={'processor-editor'}
					>
						<p>Processor Config</p>
					</div>
				}

			</div>
		)
	}
}

Editor.propTypes = {
	processor: PropTypes.object.isRequired,
	onUpdateProcessor: PropTypes.func.isRequired,
	form: PropTypes.object.isRequired
};