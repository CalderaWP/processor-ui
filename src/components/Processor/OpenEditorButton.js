import React from 'react';
import PropTypes from 'prop-types';

/**
 * Displays button to open or close processor editor
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export const OpenEditorButton = (props) => {
	return (

			<button
				style={props.style}
				className={props.className}
				onClick={props.onClick}
			>
				{!props.isOpen &&
					<span>{props.labelToOpen}</span>
				}
				{props.isOpen &&
					<span>{props.labelToClose}</span>
				}
			</button>

	)

};

/**
 * Props requirements for OpenEditorButton component
 *
 * @type {{isOpen: *, labelToOpen: shim, labelToClose: shim, onClick: *}}
 */
OpenEditorButton.propTypes = {
	isOpen: PropTypes.bool.isRequired,
	labelToOpen: PropTypes.string,
	labelToClose: PropTypes.string,
	onClick: PropTypes.func.isRequired,
};

/**
 * Default props for OpenEditorButton
 *
 * @type {{labelToOpen: string, labelToClose: string}}
 */
OpenEditorButton.defaultProps = {
	labelToOpen: 'Edit Settings',
	labelToClose: 'Close Editor',
};
