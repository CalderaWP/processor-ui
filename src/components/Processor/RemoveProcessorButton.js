import React from 'react';
import PropTypes from 'prop-types';

/**
 * Displays button to open or close processor editor
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export class  RemoveProcessorButton extends React.PureComponent {

	/**
	 * Constructor for RemoveProcessorButton
	 *
	 * @param {object} props
	 */
	constructor(props){
		super(props);
		this.state = {
			message: props.label,
			inConfirm: false
		};
		this.handleClick = this.handleClick.bind(this);
	}

	/**
	 * Handle button clicks
	 *
	 * On first click, change to confirm message,
	 * On second click, call onClick prop
	 */
	handleClick(){
		if (false === this.state.inConfirm) {
			this.setState({
				message: this.props.labelToConfirm,
				inConfirm:true
			})
		} else {
			this.props.onClick();
		}
	}

	/**
	 * Render the RemoveProcessorButton componentWithState
	 * @returns {*}
	 */
	render(){
		return (

			<button
				style={this.props.style}
				className={this.props.className}
				onClick={this.handleClick}
			>
				<span>{this.state.message}</span>
			</button>

		)
	}


};

/**
 * Props requirements for RemoveProcessorButton componentWithState
 *
 *
 * @type {{label: shim, labelToConfirm: shim, onClick: *}}
 */
RemoveProcessorButton.propTypes = {
	label: PropTypes.string,
	labelToConfirm: PropTypes.string,
	onClick: PropTypes.func.isRequired,
};

/**
 * Default props for RemoveProcessorButton
 *
 * @type {{labelToOpen: string, labelToConfirm: string}}
 */
RemoveProcessorButton.defaultProps = {
	label: 'Remove Processor',
	labelToConfirm: 'Are you sure?',
};
