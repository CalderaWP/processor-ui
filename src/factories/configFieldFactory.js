import {
	getConfigFieldValueOrDefault,
	getConfigFieldDefaultValue} from "./util";

/**
 * Factory for single processor configs
 *
 * @param {Object} configField Config fields for processor config field.
 * @param {Object} configFieldDefaults Optional. Config field defaults for processor config field.
 * @returns {{}}
 */
export const configFieldFactory = (configField, configFieldDefaults = {} ) => {
	configField.value = getConfigFieldValueOrDefault(configField);
	configField.default = getConfigFieldDefaultValue(configField);
	configFieldDefaults.default = getConfigFieldDefaultValue(configFieldDefaults);
	if( null === configField.value ){
		configField.value = configFieldDefaults.default;
	}
	return {
		...configFieldDefaults,
		...configField,
	}
};
