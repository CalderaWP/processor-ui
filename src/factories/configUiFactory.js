import * as CalderaComponents from '@caldera-labs/components';

/**
 * Given an array of field configurations, return array of React components for them
 *
 *
 * @param {Array}fieldConfigs
 * @return {Array<React.Component>}
 */
export const configUiFactory = (fieldConfigs) => {
	return CalderaComponents.factories.fieldSetFactory(fieldConfigs);
};
