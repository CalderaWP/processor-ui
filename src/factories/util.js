/**
 * Finds value property of config field or config field default
 * @param {Object} configField Config field to search for default in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldValue = (configField) => {
	return 'object'=== typeof configField && configField.hasOwnProperty('value')
		? configField.value
		: null;
};

/**
 * Finds default property of config field or config field default
 * @param {Object} configField Config field to search for default in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldDefaultValue = (configField) => {
	return 'object'=== typeof configField && configField.hasOwnProperty('default')
		? configField.default
		: null;
};

/**
 * Finds value of config field, falling back to default value
 * @param {Object} configField Config field to search for value or defualt in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldValueOrDefault = (configField ) => {
	return null !== getConfigFieldValue(configField)
		? getConfigFieldValue(configField)
		: null !== getConfigFieldDefaultValue(configField)
			? getConfigFieldDefaultValue(configField)
			: null;
};

/**
 * Convert an object (should have meaningful indexes) to a map
 * @param {Object} object Object to convert
 * @return {Map<any, any>}
 */
export const objectToMap = (object) => {
	const map = new Map();
	Object.keys(object).forEach(key => {
		map.set(key, object[key]);
	});
	return map;
}