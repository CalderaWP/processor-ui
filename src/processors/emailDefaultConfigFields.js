/**
 * The identifier for the value for the content type option that represents HTML emails
 *
 * @type {string}
 */
const contentTypeHtmlValueIdentifier = 'html';
/**
 * The default fields for email processors
 *
 * @type {{fromName: {label: string, desc: string, type: string, default: string}, fromEmail: {label: string, desc: string, type: string}, contentType: {label: string, desc: string, type: string, options: *[], default: string}, bodyPadding: {label: string, desc: string, type: string, conditionals: *[]}}}
 */
export const emailDefaultConfigFields = {
	fromName: {
		'label': 'From Name',
		'desc': 'Who the email is from',
		'type': 'input',
		'inputType': 'text',
		'default': 'Roy Sivan',
		isRequired: true,
	},
	fromEmail: {
		'label': 'From Email',
		'desc': 'Who the email is from',
		'type': 'input',
		'inputType': 'email',
		isRequired: true,
	},
	contentType: {
		'label': 'Content type',
		'desc': 'HTML or plain text?',
		'type': 'select',
		'options': [
			{
				value: contentTypeHtmlValueIdentifier,
				label: 'HTML',
			},
			{
				value: 'plain',
				label: 'Plain Text',
			}
		],
		default: contentTypeHtmlValueIdentifier
	},
	bodyPadding: {
		'label': 'Message Body Padding',
		'desc': 'How much space to add around message body',
		'type': 'number',
		'conditionals' : [
			(values) => {
				return contentTypeHtmlValueIdentifier === values.contentType;
			}
		]
	},
};

Object.keys( emailDefaultConfigFields ).forEach( key => {
	emailDefaultConfigFields[key].id = key;
});