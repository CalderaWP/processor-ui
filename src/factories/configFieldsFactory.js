import {configFieldFactory} from './configFieldFactory';
import {
	getConfigFieldValueOrDefault,
} from './util';

/**
 * Prepare a collection of config fields
 *
 * Optionally merges with default config fields
 * Optionally sets values from a Map of config values
 *
 * @param {Object} configFields Config fields for processor. Indexed by config ID.
 * @param {Object} configFieldsDefaults Optional. Config field defaults for processor. Indexed by config ID.
 * @param {Map} configValues Optional. A Map of config Values. Indexed by config ID.
 * @returns {Map}
 */
export const configFieldsFactory = (configFields, configFieldsDefaults = {},configValues = new Map()  ) => {
	const configFieldsMap = new Map(  );
	const clonedConfigFieldsDefaults = Object.assign({},configFieldsDefaults);

	/**
	 * Check if a configField's has a corresponding default config
	 *
	 * @param {String} configFieldId Id of config field to search for
	 * @returns {boolean}
	 */
	function hasConfigFieldDefault(configFieldId) {
		return clonedConfigFieldsDefaults.hasOwnProperty(configFieldId);
	}

	function getDefaultConfigObject(configFieldId) {
		return clonedConfigFieldsDefaults[configFieldId];
	}


	/**
	 * Set a config field into the Map that will be returned
	 *
	 * @param {String} configFieldId Config field ID
	 * @param {String|Number|Array} value Config field value
	 */
	function setConfigInMap(configFieldId, value) {
		let configField = {
			...configFields[configFieldId],
			ID: configFieldId,
			value: value
		};


		if( hasConfigFieldDefault(configFieldId)) {
			configField = configFieldFactory(configField,getDefaultConfigObject(configFieldId));
		}

		configFieldsMap.set(configFieldId, configField);
	}

	/**
	 * Get the value from the config field if possible
	 * @param {String} configFieldId Config field ID to search fo
	 * @returns {String|Number|Array|null} Returns null if not found
	 */
	function getFieldValue(configFieldId){
		return configValues.has( configFieldId )
			? configValues.get( configFieldId )
			: null;
	}

	//Priority for setting value:
	// fieldValues[configFieldId]
	// configField.value
	// configField.default
	// defaultConfigField.default

	//Collect config fields, adding values as needed
	Object.keys(configFields).forEach(configFieldId => {
		const value =  null !== getFieldValue(configFieldId)
			? getFieldValue(configFieldId)
			: null !== getConfigFieldValueOrDefault(configFields[configFieldId])
				? getConfigFieldValueOrDefault(configFields[configFieldId])
				: null;
		//NewProcessor to map
		setConfigInMap(configFieldId, value);
	});

	//NewProcessor in any missing fields.
	Object.keys( clonedConfigFieldsDefaults ).forEach( configFieldDefaultId => {
		if( ! configFields.hasOwnProperty(configFieldDefaultId )){
			configFieldsMap.set(configFieldDefaultId, clonedConfigFieldsDefaults[configFieldDefaultId]);
		}
	});

	return configFieldsMap;
};
