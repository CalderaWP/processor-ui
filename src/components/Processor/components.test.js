import renderer from 'react-test-renderer';
import React from 'react';
import { shallow, mount } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {OpenEditorButton} from "./OpenEditorButton";
import {RemoveProcessorButton} from "./RemoveProcessorButton";
import {Editor} from "./Editor";
import {mapKeysToIdProperty, objectToMap} from "../../factories/util";
import {emailDefaultConfigFields} from "../../processors/emailDefaultConfigFields";

import * as CalderaComponents from '@caldera-labs/components';

describe( 'Using RenderGroup', () => {
	it( 'RenderGroup snapshot', () => {
		const component = renderer.create(
			<CalderaComponents.RenderGroup configFields={[
				{
					'id': 'cf-something-select-id',
					'type': 'dropdown',
					'label': 'Content type',
					'description': 'Choose content type, default is HTML',
					options: [
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					],
					value: 'html',
					onValueChange: (newValue) =>{
						console.log(newValue)
					}
				}
			]} />
		);
		expect( component.toJSON() ).toMatchSnapshot();
	});

	it( 'RenderGroup adds our wrapper className', () => {
		const wrapper = shallow(
			<CalderaComponents.RenderGroup
				className={'cf-something-config'}
				configFields={[
				{
					'id': 'cf-something-select-id',
					'type': 'dropdown',
					'label': 'Content type',
					'description': 'Choose content type, default is HTML',
					options: [
						{
							label: 'HTML',
							value: 'html'
						},
						{
							label: 'Plain Text',
							value: 'plain'
						}
					],
					value: 'html',
					onValueChange: (newValue) =>{
						console.log(newValue)
					}
				}
			]} />
		);
		expect( wrapper.find('.cf-something-config').length ).toBe(1);
	});
});




Enzyme.configure({ adapter: new Adapter() });

describe( 'Processor components', () =>{
	function handler(){};

	describe( 'OpenEditorButton', () => {
		it( 'renders with out the non-required props', () => {
			const component = renderer.create(
				<OpenEditorButton
					isOpen={false}
					onClick={handler}
				/>
			);
			expect( component.toJSON( ) ).toMatchSnapshot();
		});

		it( 'renders with label props', () => {
			const component = renderer.create(
				<OpenEditorButton
					isOpen={false}
					onClick={handler}
					labelToOpen={'Open!'}
					labelToClose={'Close!'}
				/>
			);
			expect( component.toJSON( ) ).toMatchSnapshot();
		});

		it( 'Shows open label when open',  () =>{
			const wrapper = shallow(
				<OpenEditorButton
					isOpen={true}
					onClick={handler}
					labelToOpen={'It is closed'}
					labelToClose={'It is opened'}
				/>
			);
			expect( wrapper.find('button').text()).toEqual( 'It is opened' );
		});

		it( 'Shows closed label when closed',  () =>{
			const wrapper = shallow(
				<OpenEditorButton
					isOpen={false}
					onClick={handler}
					labelToOpen={'It is closed'}
					labelToClose={'It is opened'}
				/>
			);
			expect( wrapper.find('button').text()).toEqual( 'It is closed' );
		});

		it( 'Calls props.onClick when clicked', () => {
			let clicked = false;
			const wrapper = shallow(
				<OpenEditorButton
					onClick={() => {
						clicked = true;
					}}
					isOpen={true}

				/>
			);
			wrapper.find( 'button' ).simulate('click')
			expect( clicked).toEqual( true );
		});
	});

	describe( 'Remove Processor Button', () => {
		it( 'Renders without label props', () => {
			const component = renderer.create(
				<RemoveProcessorButton
					onClick={handler}
				/>
			);
			expect( component.toJSON( ) ).toMatchSnapshot();
		});

		it( 'Renders with label props', () => {
			const component = renderer.create(
				<RemoveProcessorButton
					onClick={handler}
					label={'R2'}
					labelToConfirm={'confirm?'}
				/>
			);
			expect( component.toJSON( ) ).toMatchSnapshot();
		});

		it( 'changes the message when clicked once', () => {
			const wrapper = shallow(
				<RemoveProcessorButton
					onClick={()=>{}}
					label={'R2'}
					labelToConfirm={'Confirm?'}
				/>
			);
			wrapper.find('button').simulate('click');
			expect(wrapper.find('button').text()).toEqual( 'Confirm?');
		});

		it( 'Does not call the onClick when clicked twice', () => {
			let clickRecived = false;
			const wrapper = shallow(
				<RemoveProcessorButton
					onClick={()=>{
						clickRecived = true;
					}}
					label={'R2'}
					labelToConfirm={'Confirm?'}
				/>
			);
			wrapper.find('button').simulate('click');
			expect(clickRecived).toEqual( false );
		});

		it( 'Calls the onClick when clicked twice', () => {
			let clickRecived = false;
			const wrapper = shallow(
				<RemoveProcessorButton
					onClick={()=>{
						clickRecived = true;
					}}
					label={'R2'}
					labelToConfirm={'Confirm?'}
				/>
			);
			wrapper.find('button').simulate('click');
			wrapper.find('button').simulate('click');
			expect(clickRecived).toEqual( true );
		});
	});

	describe( 'Editor', () => {
		const emailProcessor = {
			ID: 'p1',
			label: 'Main Email',
			type: 'email'
		};
		const processors = new Map();
		processors.set( 'p1', emailProcessor );

		const form = {
			ID: 'CF1',
			name: 'Contact',
			field: {
				firstName: {

				},
				lastName: {

				}
			}
		};

		it( 'renders with props', () => {
			const component = renderer.create(
				<Editor
					processor={emailProcessor}
					onUpdateProcessor={handler}
					form={form}
				/>
			);
			expect( component.toJSON( ) ).toMatchSnapshot();
		});

		it( 'Has chooser if no type',  () =>{
			const wrapper = shallow(
				<Editor
					processor={{
						ID: 'p122'
					}}
					onUpdateProcessor={handler}
					form={form}
				/>
			);
			expect( wrapper.find( '.processor-type-chooser').length).toBe(1);
		});

		it( 'Shows type',  () =>{
			const wrapper = shallow(
				<Editor
					processor={{
						ID: 'p122',
						type: 'redirect'
					}}
					onUpdateProcessor={handler}
					form={form}
				/>
			);
			expect( wrapper.find( '.processor-type').length).toBe(1);
			expect( wrapper.find( '.processor-type').text()).toBe('redirect');
		});

		it( 'Updates type',  () =>{
			let type = 'redirect';
			function changeHandler(updateProcessor){

				type = updateProcessor.type;
			}
			const wrapper = shallow(
				<Editor
					processor={{
						ID: 'p122',
					}}
					onUpdateProcessor={changeHandler}
					form={form}
				/>
			);
			expect( wrapper.find( '.processor-type-chooser').length).toBe(1);
			wrapper.find( '.processor-type-chooser' ).simulate('change', {target: { value : 'email'}});
			expect( type).toBe('email');
		});

	});

});