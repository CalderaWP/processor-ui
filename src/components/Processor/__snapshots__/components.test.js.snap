// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processor components Editor renders with props 1`] = `
<div
  className={undefined}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div />
  <div>
    Type: 
    <span
      className="processor-type"
    >
      email
    </span>
  </div>
</div>
`;

exports[`Processor components OpenEditorButton renders with label props 1`] = `
<button
  className={undefined}
  onClick={[Function]}
  style={undefined}
>
  <span>
    Open!
  </span>
</button>
`;

exports[`Processor components OpenEditorButton renders with out the non-required props 1`] = `
<button
  className={undefined}
  onClick={[Function]}
  style={undefined}
>
  <span>
    Edit Settings
  </span>
</button>
`;

exports[`Processor components Remove Processor Button Renders with label props 1`] = `
<button
  className={undefined}
  onClick={[Function]}
  style={undefined}
>
  <span>
    R2
  </span>
</button>
`;

exports[`Processor components Remove Processor Button Renders without label props 1`] = `
<button
  className={undefined}
  onClick={[Function]}
  style={undefined}
>
  <span>
    Remove Processor
  </span>
</button>
`;

exports[`Using RenderGroup RenderGroup snapshot 1`] = `
<div
  className=""
>
  <div
    className="caldera-config-group"
  >
    <label
      htmlFor="cf-something-select-id"
    >
      Content type
    </label>
    <input
      aria-describedby="cf-something-select-id-description"
      className="field-config"
      id="cf-something-select-id"
      onChange={[Function]}
      required={false}
      type="text"
      value="html"
    />
    <p
      className="description"
      id="cf-something-select-id-description"
    >
      Choose content type, default is HTML
    </p>
  </div>
</div>
`;
